contrib/hyperscan/matching/input_matchers/source/matcher.cc

```diff
@@ -1,91 +1,71 @@
-  flags_.reserve(config.regexes_size());
-  ids_.reserve(config.regexes_size());
-  for (const auto& regex : config.regexes()) {
-    expressions.push_back(regex.regex().c_str());
-    unsigned int flag = 0;
-    if (regex.caseless()) {
-      flag |= HS_FLAG_CASELESS;
-    }
-    if (regex.dot_all()) {
-      flag |= HS_FLAG_DOTALL;
-    }
-    if (regex.multiline()) {
-      flag |= HS_FLAG_MULTILINE;
-    }
-    if (regex.allow_empty()) {
-      flag |= HS_FLAG_ALLOWEMPTY;
-    }
-    if (regex.utf8()) {
-      flag |= HS_FLAG_UTF8;
-      if (regex.ucp()) {
-        flag |= HS_FLAG_UCP;
-      }
-    }
-    if (regex.combination()) {
-      flag |= HS_FLAG_COMBINATION;
-    }
-    if (regex.quiet()) {
-      flag |= HS_FLAG_QUIET;
-    }
-    flags_.push_back(flag);
-    ids_.push_back(regex.id());
+ScratchThreadLocal::~ScratchThreadLocal() { hs_free_scratch(scratch_); }
+
+Bound::Bound(uint64_t start, uint64_t end) : start_(start), end_(end) {}
+
+bool Bound::operator<(const Bound& other) const {
+  if (start_ == other.start_) {
+    return end_ > other.end_;
   }
+  return start_ < other.start_;
+}
 
-  hs_compile_error_t* compile_err;
-  hs_error_t err =
-      hs_compile_multi(expressions.data(), flags_.data(), ids_.data(), expressions.size(),
-                       HS_MODE_BLOCK, nullptr, &database_, &compile_err);
-  if (err != HS_SUCCESS) {
-    std::string compile_err_message(compile_err->message);
-    int compile_err_expression = compile_err->expression;
-    hs_free_compile_error(compile_err);
+Matcher::Matcher(const std::vector<const char*>& expressions,
+                 const std::vector<unsigned int>& flags, const std::vector<unsigned int>& ids,
+                 ThreadLocal::SlotAllocator& tls, bool report_start_of_matching)
+    : tls_(ThreadLocal::TypedSlot<ScratchThreadLocal>::makeUnique(tls)) {
+  ASSERT(expressions.size() == flags.size());
+  ASSERT(expressions.size() == ids.size());
 
-    if (compile_err_expression < 0) {
-      throw EnvoyException(fmt::format("unable to compile database: {}", compile_err_message));
-    } else {
-      throw EnvoyException(fmt::format("unable to compile pattern '{}': {}",
-                                       expressions.at(compile_err_expression),
-                                       compile_err_message));
+  // Compile database.
+  compile(expressions, flags, ids, &database_);
+
+  // Compile SOM database. The SOM database will report start of matching, works for replaceAll.
+  if (report_start_of_matching) {
+    std::vector<unsigned int> som_flags = flags;
+    for (unsigned int& som_flag : som_flags) {
+      som_flag = som_flag | HS_FLAG_SOM_LEFTMOST;
     }
+    compile(expressions, som_flags, ids, &som_database_);
   }
 
-  tls_->set(
-      [this](Event::Dispatcher&) { return std::make_shared<ScratchThreadLocal>(this->database_); });
+  tls_->set([this](Event::Dispatcher&) {
+    return std::make_shared<ScratchThreadLocal>(this->database_, this->som_database_);
```

Comment:
nit: `this->` not necessary -- I think that would only be needed in a template.

Reply:
Done.